app/utils/kci_test/regressions.py

```diff
@@ -0,0 +1,466 @@
+
+    return ret_val, doc_id
+
+
+def check_and_track(test_group_doc, db_options):
+    """Check previous test group report and start tracking regressions.
+
+    :param test_group_doc: The test group document we are working on.
+    :type test_group_doc: dict
+    :param conn: The database connection.
+    :param db_options: The database connection parameters.
+    :type db_options: dict
+    :return tuple The return value; and the regression document id.
+    """
+    ret_val = None
+    doc_id = None
+
+    b_get = test_group_doc.get
+    # Look for an older and as much similar as possible test group report.
+    # In case the test group report we are analyzing is FAIL and the old one
+    # is PASS, it's a new regression that we need to track.
+    spec = {
+        models.ARCHITECTURE_KEY: b_get(models.ARCHITECTURE_KEY),
+        models.BOARD_KEY: b_get(models.BOARD_KEY),
+        models.COMPILER_VERSION_EXT_KEY:
+            b_get(models.COMPILER_VERSION_EXT_KEY),
+        models.CREATED_KEY: {"$lt": b_get(models.CREATED_KEY)},
+        models.DEFCONFIG_FULL_KEY: b_get(models.DEFCONFIG_FULL_KEY),
+        models.DEFCONFIG_KEY: b_get(models.DEFCONFIG_KEY),
+        models.GIT_BRANCH_KEY: b_get(models.GIT_BRANCH_KEY),
+        models.JOB_KEY: b_get(models.JOB_KEY),
+        models.LAB_NAME_KEY: b_get(models.LAB_NAME_KEY),
+        models.NAME_KEY: b_get(models.NAME_KEY),
+    }
+
+    old_doc = utils.db.find_one3(
+        models.TEST_GROUP_COLLECTION, spec, sort=[(models.CREATED_KEY, -1)])
+
+    if old_doc:
+        database = utils.db.get_db_connection(db_options)
+        _add_test_group_data(old_doc, database)
+
+        # If previous test failed this is a number higher than 0
+        if old_doc.get('total').get('FAIL') != 0:
+            # "Old" regression case, we might have to keep track of it.
+            utils.LOG.info("Previous test group report failed, checking previous regressions")
+
+            # Check if we have old regressions first.
+            # If not, we don't track it since it's the first time we
+            # know about it.
+            prev_reg = check_prev_regression(test_group_doc, old_doc, db_options)
+
+            if all([prev_reg[0], prev_reg[1]]):
+                utils.LOG.info("Found previous regressions, keep tracking")
+                ret_val, doc_id = track_regression(
+                    test_group_doc, None, prev_reg, db_options)
+            else:
+                utils.LOG.info("No previous regressions found, not tracking")
+        # Previous test didn't fail, this is a NEW regression
+        else:
+            # New regression case.
+            utils.LOG.info("Previous test group report passed, start tracking")
+            ret_val, doc_id = track_regression(
+                test_group_doc, old_doc, (None, None), db_options)
+    else:
+        utils.LOG.info("No previous test group report found, not tracking")
+
+    return ret_val, doc_id
+
+
+def _add_test_group_data(group, database):
```

Comment:
Err, is this function not already in `app/utils/report/test.py`?  Also, why is it needed to update some test group data to while searching for regressions?

Reply:
OK sounds like something to consolidate in a follow-up PR.